packages/core/content-manager/server/src/controllers/relations.ts

```diff
@@ -233,74 +258,66 @@ export default {
         .select(`${alias}.id`)
         .getKnexQuery();
 
-      addFiltersClause(queryParams, { id: { $notIn: knexSubQuery } });
+      addFiltersClause(queryParams, {
+        // We change the operator based on whether we are looking for available or
+        // existing relations
+        id: available ? { $notIn: knexSubQuery } : { $in: knexSubQuery },
+      });
+    }
+
+    if (_q) {
+      // searching should be allowed only on mainField for permission reasons
+      const _filter = isOperatorOfType('where', query._filter) ? query._filter : '$containsi';
+      addFiltersClause(queryParams, { [mainField]: { [_filter]: _q } });
     }
 
-    // We find a page of the targeted model (relation) to display in the
-    // relation select component in the CM edit view
     const res = await strapi.entityService.findPage(
       targetUid as Common.UID.ContentType,
       queryParams
     );
 
+    if (status) {
+      // The result will contain all relations in the requested status, and we don't need to find
+      // the latest status for each.
+
+      ctx.body = {
+        ...res,
+        results: res.results.map((relation) => {
+          return {
+            ...relation,
+            status,
+          };
+        }),
+      };
+      return;
+    }
+
+    // No specific status was requested, we should find the latest available status for each relation
+    const documentMetadata = getService('document-metadata');
     ctx.body = {
       ...res,
-      results: Array.isArray(res.results) ? mapResults(res.results) : [],
+      results: await mapAsync(res.results, async (relation: RelationEntity) => {
+        const { data: documentWithLatestStatus } =
+          await documentMetadata.formatDocumentWithMetadata(targetUid, relation, {
+            availableStatus: true,
+          });
+
+        return documentWithLatestStatus;
+      }),
```

Comment:
not against this approach!, there is the option which might reduce the number of queries https://github.com/strapi/strapi/blob/v5/main/packages/core/content-manager/server/src/controllers/collection-types.ts#L135-L144

Reply:
Done in [`25ddc83` (#19681)](https://github.com/strapi/strapi/pull/19681/commits/25ddc8344408329419bd13084b76ba04b75899b4)

wdyt?

Reply:
lgtm! 🚀